Makefile / library path modifications to make M1 building work (koreader/koreader#8797)

ptrm opened the pull request:
These are mostly changes mentioned in issue koreader/koreader#8797 implemented. So far koreader builds and shows the emulator.

I used both `/usr/local/opt` and `/opt/homebrew/` Homebrew prefixes, as the Apple Silicon Homebrew path has been only recently established to `/opt/homebrew` (https://apple.stackexchange.com/a/410829), though some users may have older brew installation with `/usr/local/opt` prefix (as opposed to `/usr/local/` without `opt` for Intel brew).

<img width="835" alt="image" src="https://user-images.githubusercontent.com/366258/210275163-3c32dc76-739d-473e-a335-5916a441e9f7.png">

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader-base/1570)
<!-- Reviewable:end -->

Frenzie commented on ffi/SDL2_0.lua:
```diff
@@ -21,6 +21,8 @@ require("ffi/linux_input_h")
 -----------------------------------------------------------------
 
 local SDL = util.ffiLoadCandidates{
+    "/usr/local/opt/lib/libSDL2.dylib",
+    "/opt/homebrew/lib/libSDL2.dylib",
```
This should go lower since SDL2 is the default. See <http://luajit.org/ext_ffi_api.html#ffi_load>.

>  On POSIX systems, if the name contains no dot, the extension .so is appended. Also, the lib prefix is prepended if necessary. So ffi.load("z") looks for "libz.so" in the default shared library search path.
>
> On Windows systems, if the name contains no dot, the extension .dll is appended. So ffi.load("ws2_32") looks for "ws2_32.dll" in the default DLL search path.

Frenzie commented on ffi/util.lua:
```diff
@@ -619,7 +619,7 @@ function util.haveSDL2()
         local candidates
         if jit.os == "OSX" then
-            candidates = {"libs/libSDL2.dylib", "SDL2"}
+            candidates = {"/opt/homebrew/lib/libSDL2.dylib", "SDL2"}
```
It looks like this should be added, not replaced.

Frenzie commented on thirdparty/glib/CMakeLists.txt:
```diff
@@ -69,8 +69,8 @@ endif()
 else()
-    set(CFLAGS "-I/usr/local/opt/gettext/include ${CFLAGS}")
-    set(LDFLAGS "${LDFLAGS} -L/usr/local/opt/gettext/lib -L${LIBICONV_DIR}/lib -L${ZLIB_DIR}/lib -L${LIBFFI_DIR}/.libs -L${BINARY_DIR}/gmodule/.libs")
+    set(CFLAGS "-I/usr/local/opt/gettext/include -I/opt/homebrew/include ${CFLAGS}")
+    set(LDFLAGS "${LDFLAGS} -L/usr/local/opt/gettext/lib -L/opt/homebrew/lib -L${LIBICONV_DIR}/lib -L${ZLIB_DIR}/lib -L${LIBFFI_DIR}/.libs -L${BINARY_DIR}/gmodule/.libs")
```
I don't like this hardcoding.

ptrm replied:
You mean `/opt/homebrew/lib` should get a variable and become ${HOMEBREW_PREFIX}/lib?

Also is it a good idea to check for brew binary in `Makefile.defs` in case of Darwin host and set the `HOMEBREW_PREFIX` from `$(shell brew --prefix)`? And could this be passed to the lua files somehow? This way custom prefixes set during brew installation could be covered, though probably people who bother to set them know how to fix their builds later on ;)

Frenzie replied:
Definitely that, but I'd also prefer to be able to plug in e.g. Nix or gentoo-prefix with little effort if reasonably possible. `HOMEBREW_PREFIX` might allow for that but then the name (at least further down at this point) should really just be something like `MAC_LIBRARY_PREFIX`. But you could always write `MAC_LIBRARY_PREFIX := HOMEBREW_PREFIX` up above in defs or something so it's self-documenting without comments.

NiLuJe replied:
IIRC, there's a runtime homebrew command that will actually spit out the right prefix for the current system ;).

ptrm replied:
Yup, there's `brew --prefix`. I'll try to umharfcode the paths around evening CET.

What about the paths in lua part, can / should they somehow be based on `brew prefix` as well?

Frenzie replied:
Might it be more of a system configuration issue if it doesn't look in `/opt/homebrew/lib` by itself? Put another way, it doesn't seem to be an issue for me on Intel Mac.

Frenzie replied:
Basically up above `SDL2` → `libSDL2.dylib` is part of LuaJIT, and then it just checks a bunch of paths for `libSDL2.dylib`. The .so stuff is because `libSDL2-2.0.so` isn't a sane name like `libSDL2.so`. ;-P

NiLuJe replied:
Yeah, that might be better left to whatever's macOS version of LD_LIBRARY_PATH is (I can't recall ;p).

ptrm replied:
@Frenzie Homebrew keeps linked libs in /usr/local/lib for Intel, which is a common path for tools to search (LD*_PATH is considered insecure for some reason on Darwins). /opt/homebrew was added for apple m1. Here is the descrpition, along with the mention about [ffi ignoring /opt/homebrew/lib until 1.15.2](https://earthly.dev/blog/homebrew-on-m1/#update-third-party-code)? 

>  The problem I encountered was that ffi did not add /opt/homebrew/lib/ to its list of search paths until [version 1.15.2](https://github.com/ffi/ffi/blob/master/CHANGELOG.md#1152--2021-06-16). We were running v1.15.0 ,so every time I tried to run the blog I kept getting no such file or directory errors because ffi wasn’t updated to check /opt/homebrew/lib

I think this might be the reason on the lua side.

NiLuJe replied:
LuaJIT doesn't use libffi ;).

ptrm replied:
To double check: you have no objections about using `brew --prefix` for MACOS on the Makefile part?

NiLuJe replied:
`ffi.load` is a thin wrapper around `dlopen`, which will do whatever the system loader is configured (i.e., honor LD_LIBRARY_PATH in the case of GNU ld).

ptrm replied:
> `ffi.load` is a thin wrapper around `dlopen`, which will do whatever the system loader is configured (i.e., honor LD_LIBRARY_PATH in the case of GNU ld).

Agreed for linux, but as [linked before](https://medium.com/macos-is-not-linux-and-other-nix-reflections/d-o-y-ou-ld-library-path-you-6ab0a6135a33), on Darwins you don't rely much on LD_LIBRARY_PATHS

NiLuJe replied:
I don't care about Apple's inane design/"security" choice. We *heavily* rely on custom search paths (be it via env vars of DT_RPATH entries baked in binaries) ;).

ptrm replied:
Ok, but what of the change in ffi that picks up the `/opt/homebrew/lib` path on its own? I could not find which ffi version is used for koreader, but would it add less complications to the run process to just bump ffi?

ptrm replied:
Aside from the fact that I linked Ruby-FFI issues XD

Frenzie replied:
> Homebrew keeps linked libs in /usr/local/lib for Intel, which is a common path for tools to search

Judging by `man dyld` you should try something like `export DYLD_FALLBACK_LIBRARY_PATH="${DYLD_FALLBACK_LIBRARY_PATH}:/opt/homebrew/lib"`.

ptrm replied:
How would you suggest to provide the SDL path to the emulator then? I tried removing the full-path candidates and running  `DYLD_LIBRARY_PATH=/opt/homebrew/lib ./kodev run`, also `export  DYLD_LIBRARY_PATH=/opt/homebrew/lib` and then run, but got 

```
ffi.load: /opt/homebrew/lib/libSDL2.dylib
ffi.load: SDL2
ffi.load (warning): dlopen(libSDL2.dylib, 0x0005): tried: 'libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache), 'libSDL2.dylib' (no such file), '/usr/local/lib/libSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache)
ffi.load: libSDL2-2.0.so
ffi.load (warning): dlopen(libSDL2-2.0.so, 0x0005): tried: 'libSDL2-2.0.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2-2.0.so' (no such file), '/usr/lib/libSDL2-2.0.so' (no such file, not in dyld cache), 'libSDL2-2.0.so' (no such file), '/usr/local/lib/libSDL2-2.0.so' (no such file), '/usr/lib/libSDL2-2.0.so' (no such file, not in dyld cache)
ffi.load: libSDL2-2.0.so.0
ffi.load (warning): dlopen(libSDL2-2.0.so.0, 0x0005): tried: 'libSDL2-2.0.so.0' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2-2.0.so.0' (no such file), '/usr/lib/libSDL2-2.0.so.0' (no such file, not in dyld cache), 'libSDL2-2.0.so.0' (no such file), '/usr/local/lib/libSDL2-2.0.so.0' (no such file), '/usr/lib/libSDL2-2.0.so.0' (no such file, not in dyld cache)
./luajit: ./ffi/SDL2_0.lua:34: attempt to index local 'SDL' (a boolean value)
```

Frenzie replied:
> I could not find which ffi version is used for koreader, but wouldn't it add less complications to the run process to just bump ffi?

None. libffi is like LuaJIT FFI for regular Lua. ;-)

ptrm replied:
the export DYLD_FALLBACK_LIBRARY_PATH also fails to help find libsdl :/

Frenzie replied:
> Imposing GNU tools compliance on macs is inane for me, in other words

It's not "GNU tools compliance". It's "hardcoding Homebrew as if Homebrew were the one and only thing of its type that exists is incredibly dumb and I'm nowhere close to accepting such idiocy just yet since at this point I'm not convinced saner avenues have been sufficiently explored."

Frenzie replied:
Put another way, the Lua changes are conceptually distinct from the Makefile changes.

ptrm replied:
> > Imposing GNU tools compliance on macs is inane for me, in other words
> 
> It's not "GNU tools compliance". It's "hardcoding Homebrew as if Homebrew were the one and only thing of its type that exists is incredibly dumb and I'm nowhere close to accepting such idiocy just yet since at this point I'm not convinced saner avenues have been sufficiently explored."

man dyld says:
```
 DYLD_FALLBACK_LIBRARY_PATH
              This is a colon separated list of directories that contain
              libraries.  If a dylib is not found at its install  path, dyld
              uses this as a list of directories to search for the dylib.  By
              default, it is set to /usr/local/lib:/usr/lib.
```

On my Ventura all the *_LIBRARY_PATHs I checked are empty. Maybe it's just another time Apple threw in a BSD tool, changed it but didn't bother about documenting in the the manpages, if anywhere?

The only solution I would oppose would be setting global _LIBRARY_ variables in macos, but passing it in a working way (and telling in the build.md to set it to `brew --path` or whatever one uses) looks also better to me than hardcoding the paths. I just have no idea yet how to pass it or make it work ^^

ptrm replied:
So for clarity, after removing all homebrew paths from lua files, the env variables don't seem to do much:

with variables emptied:
```
➜  koreader git:(master) ✗ export DYLD_FALLBACK_LIBRARY_PATH=''        
➜  koreader git:(master) ✗ export DYLD_FALLBACK_LIBRARY_PATH=''
➜  koreader git:(master) ✗ export LD_LIBRARY_PATH=''                  
➜  koreader git:(master) ✗ export DYLD_LIBRARY_PATH=''                
➜  koreader git:(master) ✗ ./kodev run                                    
...
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 01/03/23-22:03:46
 [*] Version: v2022.11-73-gf0b007bb_2023-01-01

ffi.load: libs/libutf8proc.2.dylib
ffi.load: blitbuffer
ffi.load (assisted searchpath): ./libs/libblitbuffer.so
ffi.load: libs/libSDL2.dylib
ffi.load (warning): dlopen(libs/libSDL2.dylib, 0x0005): tried: 'libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibs/libSDL2.dylib' (no such file), '/usr/lib/libs/libSDL2.dylib' (no such file, not in dyld cache), 'libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file)
ffi.load: SDL2
ffi.load (warning): dlopen(libSDL2.dylib, 0x0005): tried: 'libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache), 'libSDL2.dylib' (no such file), '/usr/local/lib/libSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache)
SDL2 not loaded:        ./setupkoenv.lua:27: Not able to load dynamic library: SDL2
./luajit: frontend/device.lua:58: Could not find hardware abstraction for this platform. If you are trying to run the emulator, please ensure SDL is installed.
stack traceback:
        [C]: in function 'error'
        frontend/device.lua:58: in function 'probeDevice'
        frontend/device.lua:61: in main chunk
        [C]: in function 'require'
        ./reader.lua:150: in main chunk
```

with paths provided:
```
➜  koreader git:(master) ✗ export DYLD_FALLBACK_LIBRARY_PATH=/opt/homebrew/lib
➜  koreader git:(master) ✗ export LD_LIBRARY_PATH=/opt/homebrew/lib           
➜  koreader git:(master) ✗ export DYLD_LIBRARY_PATH=/opt/homebrew/lib         
➜  koreader git:(master) ✗ ./kodev run     
...
ffi.load: libs/libSDL2.dylib
ffi.load (warning): dlopen(libs/libSDL2.dylib, 0x0005): tried: 'libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibs/libSDL2.dylib' (no such file), '/usr/lib/libs/libSDL2.dylib' (no such file, not in dyld cache), 'libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file)
ffi.load: SDL2
ffi.load (warning): dlopen(libSDL2.dylib, 0x0005): tried: 'libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache), 'libSDL2.dylib' (no such file), '/usr/local/lib/libSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache)
SDL2 not loaded:        ./setupkoenv.lua:27: Not able to load dynamic library: SDL2
```

NiLuJe replied:
otool (or similar ldd-like tools) *probably* have debug knobs to see what the loader is actually doing.

NiLuJe replied:
At the very least possibly DYLD_PRINT_LIBRARIES=1 (c.f., https://stackoverflow.com/a/51504440)

NiLuJe replied:
And apparently the hardened runtime disable dyld env vars or... something; but I don't recall how forceful XCode is about that.

NiLuJe replied:
But SIP probably means if you`exec` from an SIP app, you're essentially fucked? Yay, Apple.

NiLuJe replied:
So, yeah, seems like a giant "fuck you, FOSS", from Apple. c.f., https://github.com/Homebrew/brew/issues/13481 for more context.

TL;DR: Hardcoded-search paths it is. `>_<".`

Frenzie replied:
I think you still meant to base these ones on brew --prefix or some such as well? :-)

NiLuJe replied:
Yes ;).

ptrm replied:
Ok, just pushed a commit with variables instead of hardcoded brew path ;)

Frenzie reviewed (commented):
Thanks! Looks very promising.

ptrm commented on ffi/util.lua:
```diff
@@ -60,10 +60,27 @@ int WideCharToMultiByte(
 );
 ]]
 
+local getlibprefix = function()
```
So that didn't turn out as a oneliner. `os.popen(...) or ""` was throwing errors, and as I found in other parts of the code that `os.popen` has to be taken care of as a `fopen`.

NiLuJe reviewed (commented):
(goddamnit GH review UI sucks)

NiLuJe reviewed (changes requested):
> Reviewed 3 of 5 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1570)* status: all files reviewed, 14 unresolved discussions (waiting on @Frenzie and @ptrm)

___
*[`Makefile.defs` line 11 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIUB8MEaOmB78mrGMW:-NLIUB8MEaOmB78mrGMX:bid3rjj) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/Makefile.defs#L11)):*
> ```
> 
> UNAME:=$(shell uname -s)
> UNAME_ARCH:=$(shell uname -m)
> ```

nit: I'd rename that to `HOST_ARCH` (and I'm mildly intrigued by the fact that we don't already have that stored somewhere?).

EDIT: Oh, yeah, we have it, but we use it for the actual march flag, meh :/. Nevermind that one, then.
___
*[`Makefile.defs` line 194 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIUMz_DW3leNtKQfMc:-NLIUMzaEsf5ao90Ah2F:bavwy3b) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/Makefile.defs#L194)):*
> ```
> 	export DARWINHOST=1
> 	ifneq (,$(filter $(UNAME_ARCH), arm64))
> 		export DARWINM1HOST=1
> ```

nit: `DARWIN_AARCH64_HOST` (extra columns are free!)
___
*[`Makefile.defs` line 199 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIUmOlCkuD0CMnFeHG:-NLIUmOlCkuD0CMnFeHH:b-5en96r) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/Makefile.defs#L199)):*
> ```
> 		export DARWIN=1
> 		ifdef DARWINM1HOST
> 			export DARWIN_M1=1
> ```

nit: `DARWIN_AARCH64`
___
*[`Makefile.defs` line 511 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIUx_MBWMobFo9LRIU:-NLIUx_MBWMobFo9LRIV:b-ot20iz) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/Makefile.defs#L511)):*
> ```
> 
> ifdef DARWINM1HOST
> 	HOST_ARCH:=-mcpu=apple-m1
> ```

And a comment that M1 is the lowest common denominator for Apple AARCH64 silicon for future archeologists ;).
___
*[`ffi/SDL2_0.lua` line 28 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIVLB76h4AIp9KzKfq:-NLIVLB76h4AIp9KzKfr:bls3uqj) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/ffi/SDL2_0.lua#L28)):*
> ```Lua
>     "libSDL2-2.0.so",
>     "libSDL2-2.0.so.0",
>     util.KO_DYLD_PREFIX .. "/lib/libSDL2.dylib",
> ```

Mildly mystified why this one doesn't have the `libs/libSDL2.dylib` variant found in `util`?
___
*[`ffi/util.lua` line 63 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIVgns-7W9Bp7nxge_:-NLIVgns-7W9Bp7nxgea:b-21of0s) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/ffi/util.lua#L63)):*
> ```Lua
> ]]
> 
> local getlibprefix = function()
> ```

Missing a `jit.os == "OSX"` branch, so we don't try to run brew everywhere.
___
*[`ffi/util.lua` line 64 at r3](https://reviewable.io/reviews/koreader/koreader-base/1570#-NLIVVRWDieJY3nDnr5X:-NLIVVRWDieJY3nDnr5Y:bad9woj) ([raw file](https://github.com/koreader/koreader-base/blob/8b7a8a98c86bf31f9ba987e6a207ba52759f5b01/ffi/util.lua#L64)):*
> ```Lua
> 
> local getlibprefix = function()
>     -- Apple M1 homebrew installs libraries outside of default searchpaths,
> ```

nit: AArch64 ;)
___
*[`thirdparty/glib/CMakeLists.txt` line 73 at r1](https://reviewable.io/reviews/koreader/koreader-base/1570#-NKoCAkwhzN1rLG7OWc1-r1-73:-NLIWDMe5fLXSgISn6rD:bre4a1n) ([raw file](https://github.com/koreader/koreader-base/blob/90f641c722048affe44360a0dbadb1c512dfd874/thirdparty/glib/CMakeLists.txt#L73)):*
<details><summary><i>Previously, Frenzie (Frans de Jonge) wrote…</i></summary><blockquote>

I think you still meant to base these ones on brew --prefix or some such as well? :-)
</blockquote></details>

These are still hard-coded ;).


<!-- Sent from Reviewable.io -->

NiLuJe commented:
My GH "review" comments may be outdated, I had once again forgotten to hit the big green button.

The Reviewable stuff is current ;).

ptrm commented on Makefile.defs:
```diff
@@ -189,8 +190,14 @@ endif
 ifneq (,$(filter $(UNAME), Darwin))
 	export DARWINHOST=1
+	ifneq (,$(filter $(UNAME_ARCH), arm64))
+		export DARWINM1HOST=1
```
True, not without a reason being called Apple Sillicon ;> Will fix

ptrm commented on ffi/util.lua:
```diff
@@ -60,10 +60,27 @@ int WideCharToMultiByte(
+local getlibprefix = function()
+    -- Apple M1 homebrew installs libraries outside of default searchpaths,
+    -- and dyld environment variables are sip-protected on MacOS, cf. https://github.com/Homebrew/brew/issues/13481#issuecomment-1181592842
+    local libprefix = os.getenv("KO_DYLD_PREFIX");
```
Sorry, I'm a js guy :v Btw I see linter remote checks. Should I get a warning for this locally (in a properly configured editor, that is) based on some definitions in the source?

Frenzie replied:
We don't currently have anything that checks for this because it's a style thing and not really an issue otherwise. Also (almost) no one adds redundant semicolons in Lua. :-P

NiLuJe commented:
And I just noticed that I mixed aarch64 w/ arm64 in my renaming requests; aarch64 is technically more correct, but arm64 might be easier on the eyes? ;).

NiLuJe commented on ffi/util.lua:
```diff
@@ -60,10 +60,27 @@ int WideCharToMultiByte(
+            end
+    end
+
+    return libprefix;
```
nit: trailing semi-colon

Frenzie replied:
Indeed, better add some more:
```suggestion
    return libprefix;;;;;;;;;;
```

NiLuJe reviewed (changes requested):
LGTM besides that very small nit ;).

ptrm commented:
Removed the semicolon, I'm blind to them at EOL ;) Also added a comment proposed in Reviewable about apple-m1 being lowest common denominator for Apple silicon chips.

Frenzie commented on ffi/util.lua:
```diff
@@ -60,10 +60,27 @@ int WideCharToMultiByte(
 
 local util = {}
 
+util.KO_DYLD_PREFIX = getlibprefix()
```
This shouldn't be called except on OS X, so probably on line 654 below.

ptrm replied:
Yeah, thanks for the unbreaking commit elsewhere, I didn't consider this might throw an exception ;)

Frenzie replied:
Good thing the CI caught it. ;-)
